### DISCUSSION TITLE
Incorporating best practices for Accessibility in the documentation

### DISCUSSION THREAD

**Discussion by Malvoz** [Leaflet/Leaflet#8006] [Ideas]
[Accessibility efforts](https://github.com/Leaflet/Leaflet/issues?q=label%3Aaccessibility) for Leaflet can provide a more accessible baseline for the different features/components, and the map as a whole.

But sometimes it's the responsibility of developers who implement Leaflet maps to take extra steps to ensure their maps are presented in a way that they can be used and make sense to people of all abilites.

This discussion is about the things that Leaflet can't do for developers out-of-the-box, but rather how updates/additions to docs can help developers create more accessible maps. Also things that can be improved to ensure the site is usable for keyboard users and users with assistive technologies.

Areas that can be improved:

- [Accessibility best practices and tips in tutorials](https://github.com/Leaflet/Leaflet/discussions/8006#discussioncomment-2190730)
- [Dedicated tutorial/guideline page for accessibility](https://github.com/Leaflet/Leaflet/discussions/8006#discussioncomment-2190734)
- [Advisory information in the reference docs](https://github.com/Leaflet/Leaflet/discussions/8006#discussioncomment-2190747)
- [Plugin Authoring Guide: Content Accessibility](https://github.com/Leaflet/Leaflet/discussions/8006#discussioncomment-2199260)
- [Misc / General site issues](https://github.com/Leaflet/Leaflet/discussions/8006#discussioncomment-2228800)

Feel free to discuss each, and give suggestions.

**Comment by Malvoz** [comment]
## Accessibility best practices and tips in tutorials

For the existing [tutorials](https://leafletjs.com/examples.html), check:

1. If they generally follow accessibility guidelines and best practices (so that everyone can use them, and they don't introduce accessibility issues to people's sites if they copy & paste), e.g. that they use [valid HTML](https://validator.w3.org/) and appropriate [semantics](https://developer.mozilla.org/en-US/docs/Glossary/Semantics) (also run [link checks](https://validator.w3.org/checklink) for good measure);
2. If there are opportunities to incorporate specific accessibility tips

---

### Reviews of tutorials (in progress)

_If you find an unchecked item in the reviews of these tutorials, feel free to send a PR to fix it._

<details>
<summary>Quick Start Guide</summary><br>

The [Quick Start Guide](https://leafletjs.com/examples/quick-start/):

- [ ] Markers should have descriptive `title` or `alt`

</details>

<details>
<summary>Leaflet on mobile</summary><br>

The [Leaflet on mobile](https://leafletjs.com/examples/mobile/) tutorial:

- [ ] Shouldn't disable user-scaling in the recommended meta tag (see https://github.com/Leaflet/Leaflet/issues/7342#issuecomment-1023036026)

</details>

<details>
<summary>Markers With Custom Icons</summary><br>

The [Markers With Custom Icons](https://leafletjs.com/examples/custom-icons/) tutorial:

- [ ] The icons should have an [_accessible name_](https://www.w3.org/TR/accname/#dfn-accessible-name)/[_accessible description_](https://www.w3.org/TR/accname/#dfn-accessible-description), for example "green leaf", "red leaf", "yellow leaf" e.g. via `title` or `alt` attributes

</details>

<details>
<summary>Accessible Maps Guide ✔️</summary><br>

The [Accessible Maps Guide](https://leafletjs.com/examples/accessibility/):

No changes needed.

</details>

<details>
<summary>Using GeoJSON with Leaflet</summary><br>

The [Using GeoJSON with Leaflet](https://leafletjs.com/examples/geojson/) tutorial:

- [ ] Marker should have a descriptive `title` or `alt`
- [ ] Vectors should be keyboard focusable and have [_accessible name_](https://www.w3.org/TR/accname/#dfn-accessible-name)/[_accessible description_](https://www.w3.org/TR/accname/#dfn-accessible-description)
     (blocked on https://github.com/Leaflet/Leaflet/issues/7822)

</details>

<details>
<summary>Interactive Choropleth Map</summary><br>
The <a href="https://leafletjs.com/examples/choropleth/">Interactive Choropleth Map</a> tutorial:

- [ ] A potential accessibility tip is to use ARIA, such that the info is announced when the mouse hovers over the states:

  [`aria-live="assertive"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-live) on `<div class="info leaflet-control">` (you may have to unmute the video):

  https://user-images.githubusercontent.com/26493779/154323657-9fd83cb4-9c0d-4a69-960e-e9bc67fe8c62.mp4

  (However, for vectors in general, https://github.com/Leaflet/Leaflet/issues/7822 would be a better long-term solution.)

  - [ ] In the video above, the "/ mi<sup>2</sup>" (`/ mi<sup>2</sup>`) is announced as <q>slash, me, group 2</q>. This could be improved with something like `<span aria-label="per square mile">/ mi<sup>2</sup></span>`.

- [ ] Vectors should be keyboard focusable and have [_accessible name_](https://www.w3.org/TR/accname/#dfn-accessible-name)/[_accessible description_](https://www.w3.org/TR/accname/#dfn-accessible-description)
     (blocked on https://github.com/Leaflet/Leaflet/issues/7822)

</details>

<details>
<summary>Layer Groups and Layers Control</summary><br>

The [Layer Groups and Layers Control](https://leafletjs.com/examples/layers-control/) tutorial:

- [ ] Markers should have descriptive `title` or `alt`

</details>

<details>
<summary>Zoom levels</summary><br>

The [Zoom levels](https://leafletjs.com/examples/zoom-levels/) tutorial:

{TODO}

</details>

<details>
<summary>Non-geographical maps</summary><br>

The [Non-geographical maps](https://leafletjs.com/examples/crs-simple/crs-simple.html) tutorial:

- [ ] Markers should have descriptive `title` or `alt`

</details>

<details>
<summary>WMS and TMS ✔️</summary><br>

The [WMS and TMS](https://leafletjs.com/examples/wms/wms.html) tutorial:

No apparent changes needed.

</details>

<details>
<summary>Working with map panes</summary><br>

The [Working with map panes](https://leafletjs.com/examples/map-panes/) tutorial:

- [ ] Vectors should be keyboard focusable and have [_accessible name_](https://www.w3.org/TR/accname/#dfn-accessible-name)/[_accessible description_](https://www.w3.org/TR/accname/#dfn-accessible-description)
     (blocked on https://github.com/Leaflet/Leaflet/issues/7822)

</details>

<details>
<summary>Overlays</summary><br>

The [Overlays](https://leafletjs.com/examples/video-overlay/) tutorial:

- [ ] Vectors should be keyboard focusable and have [_accessible name_](https://www.w3.org/TR/accname/#dfn-accessible-name)/[_accessible description_](https://www.w3.org/TR/accname/#dfn-accessible-description)
     (blocked on https://github.com/Leaflet/Leaflet/issues/7822)
- [x] The custom video controls to Play/Pause in [A bit of control over the video](https://leafletjs.com/examples/video-overlay/#a-bit-of-control-over-the-video) should have descriptive `title`/`aria-label`s.
  <s>https://github.com/Leaflet/Leaflet/pull/8017</s>  
  https://github.com/Leaflet/Leaflet/pull/8090

</details>

<details>
<summary>Extending Leaflet: Class Theory</summary><br>

- [ ] Diagram is not accessible to screen reader users (mentioned in https://github.com/Leaflet/Leaflet/issues/8086#issuecomment-1079741954)

</details>

<details>
<summary>Extending Leaflet: Layers</summary><br>

{TODO}

</details>

<details>
<summary>Extending Leaflet: Handlers and Controls</summary><br>

{TODO}

</details>

**Comment by Malvoz** [comment]
## Dedicated tutorial/guideline page for accessibility

_Feel free to discuss these items, or suggest new additions to the [Guide to basic Leaflet accessibility](https://leafletjs.com/examples/accessibility/)!_


  
  - [x] Always set a unique and descriptive [`alt`](https://leafletjs.com/reference.html#marker-alt) or [`title`](https://leafletjs.com/reference.html#marker-title) for markers.
        (Fixed in: https://github.com/Leaflet/Leaflet/pull/8081)

- [x] `divIcon` accessibility (related: https://github.com/Leaflet/Leaflet/issues/3400)
       (Fixed in: https://github.com/Leaflet/Leaflet/pull/8081)

- [ ] Always set a unique and descriptive `desc` or `title` for vectors.
      (blocked on https://github.com/Leaflet/Leaflet/pull/8251)

- [ ] Describe single maps and/or make multiple maps in a document _discernible_ to screen reader users.
     (relates to https://github.com/Leaflet/Leaflet/issues/7193, which is intended to make maps _recognizable_ to screen reader users by default.)<br>
     A unique and descriptive `aria-label`, `aria-labelledby`, `aria-description`, `aria-describedby`, `title`, `<figcaption>` in `<figure>`, and headings in `<section>` for each map may all be potential ways to achieve this.
- [ ] Mention that _standalone_ popups and tooltips can be problematic?<br>
      The problems are:
     - for keyboard users: Once popups/tooltips are closed, the user can't open them (without reloading the page).
     - for screen reader users: Missing contextual information for the popups/tooltips (that would otherwise be provided via e.g. marker `alt` or `title`).
- [ ] Provide guidance on when to use tooltips vs popups:
     - Generally, if interactive content is used (links, buttons, forms) use a popup!
       This is to align the user's expectations for Leaflet's tooltips with browsers' native tooltip behaviors (as they can't have interactive content) and [ARIA tooltip best practices](https://www.w3.org/WAI/ARIA/apg/patterns/tooltip/), but also because only [popups should recieve focus when opened](https://github.com/Leaflet/Leaflet/issues/8115) (in which case the user can easily tab to the content within).
     - Use tooltips if there is a need to show information on hover (some people may want to compromise though: https://github.com/Leaflet/Leaflet/issues/60?)
      
- [x] [_Purely decorative_](https://www.w3.org/TR/WCAG21/#dfn-pure-decoration) maps.
        (Fixed in: https://github.com/Leaflet/Leaflet/pull/8081)

**Comment by Malvoz** [comment]
## Advisory information in the reference docs

_Feel free to discuss these items, or suggest new additions to this list!_

- [ ] For each `keyboard` option ([#map-keyboard](https://leafletjs.com/reference.html#map-keyboard), [#marker-keyboard](https://leafletjs.com/reference.html#marker-keyboard)) add "Useful for accessibility" with a link to the [Preserve useful defaults](https://leafletjs.com/examples/accessibility/#preserve-useful-defaults) guideline.
- [ ] Add "Useful for accessibility" to description for [`autoPanOnFocus`](https://leafletjs.com/reference.html#marker-autopanonfocus).
- [ ] <s>Changing the default (`true`) of [#control-layers-collapsed](https://leafletjs.com/reference.html#control-layers-collapsed) results in more <kbd>Tab</kbd>s for keyboard-only users (or more swipe gestures for screen reader users on touch devices).</s>
  (<ins>This behavior has changed, but is likely to be brought back per https://github.com/Leaflet/Leaflet/issues/7191</ins>)

**Comment by Malvoz** [comment]
## Plugin Authoring Guide: Content Accessibility

_Feel free to discuss and suggest improvements to the Plugin Authoring Guide's [Content Accessibility](https://github.com/Leaflet/Leaflet/blob/main/PLUGIN-GUIDE.md#content-accessibility) section here._

**Reply by Malvoz** [reply]
Relevant discussion in https://github.com/Leaflet/Leaflet/pull/7277#pullrequestreview-805503502.

In that discussion @Falke-Design expressed that the authoritative ARIA (etc.) documents can be hard to read for those who are new to them.

MDN has some accessibility material that may be easier to read, for example:
- https://developer.mozilla.org/en-US/docs/Learn/Accessibility
- https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML
- https://developer.mozilla.org/en-US/docs/Learn/Accessibility/What_is_accessibility
- https://developer.mozilla.org/en-US/docs/Web/Accessibility

Perhaps there can be 2 sections of links instead of the current section "Learn more about web accessibility". Where "Learn more about web accessibility" would have beginner-friendly material such as MDN, and a second section for the authoritative documents?

Will think more about this.

**Reply by Malvoz** [reply]
Under [Accessibility testing](https://github.com/Leaflet/Leaflet/blob/main/PLUGIN-GUIDE.md#accessibility-testing), add headings "Automated testing" and "Manual testing". Mostly to separate the two into sections, but also to highlight the manual testing part. I think it's really important and easy for developers to do manual testing.

**Reply by Falke-Design** [reply]
+1 I think this improvements are good

**Reply by Malvoz** [reply]
- [ ] Try to shorten the section down
- [ ] Following should have links: https://github.com/Leaflet/Leaflet/blob/a3f1ba930536c036bfd5890dbae5fb18e8ee82d8/PLUGIN-GUIDE.md?plain=1#L161 but maybe it's better to link to the [Test your maps](https://leafletjs.com/examples/accessibility/#test-your-maps) guideline

**Comment by Malvoz** [comment]
## Misc / General site issues

_Feel free to discuss these items, suggest new additions to this list, or send PR(s) to fix them!_

- [ ] https://github.com/Leaflet/Leaflet/issues/7537 (https://github.com/Leaflet/Leaflet/pull/7900)
- [x] https://github.com/Leaflet/Leaflet/issues/7899 (https://github.com/Leaflet/Leaflet/pull/7901)
- [x] https://github.com/Leaflet/Leaflet/pull/7948
- [ ] Accordions in the Leaflet API reference are not keyboard accessible. See https://github.com/Leaflet/Leafdoc/issues/61 for more info.
- [ ] Blog posts semantics: Replace [`<div class="clearfix">`](https://github.com/Leaflet/Leaflet/blob/92906fdbcfe19ee84145f1045b96b94f22182f8f/docs/blog.md?plain=1#L14) with `<article class="clearfix">`?
- [ ] Remove the [**duplicate** links to tutorials from the _thumbnails_](https://leafletjs.com/examples.html) from the focus order using `tabindex="-1"`, and hide from screen readers using `aria-hidden="true"`.
- [ ] Hide the ← (`&larr;`) character using `aria-hidden` in the [`tutorials-back` link](https://github.com/Leaflet/Leaflet/blob/7a2719d49e1206975523f99ead3dc1d46fc36d2d/docs/_layouts/tutorial_v2.html#L7).
- [ ] Remove viewport meta tag's [`user-scalable`](https://github.com/Leaflet/Leaflet/search?q=user-scalable) everywhere, per [this warning](https://developer.mozilla.org/en-US/docs/Web/HTML/Viewport_meta_tag#:~:text=Usage%20of%20user%2Dscalable%3Dno%20can%20cause%20accessibility%20issues) (and also because many browsers no longer support it).
